File: 8/11/3.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;





int solution(int m, int n, vector<vector<int> > puddles) {
    int answer = 0;




    int methods[101][101];

    for(int j = 1; j <= n; j++){
        for(int i = 1; i<= m; i++){
            methods[j][i] =1;

        }

    }


    for(int i = 0; i < puddles.size(); i++){
        methods[puddles[i][1]][puddles[i][0]] = 0; 

        if(puddles[i][0]==1){
            for(int j = puddles[i][1]; j<=n; j++){
                methods[j][1] = 0;

            }
        }
        if(puddles[i][1]==1){
            for(int j = puddles[i][0]; j<=m; j++){
                methods[1][j] = 0;

            }

        }

    }

    for(int i = 2; i<= m;i++){
        for(int j = 2; j<= n; j++){
            if(methods[j][i]!=0){
                methods[j][i] = (methods[j-1][i]+methods[j][i-1])%1000000007;

            }
            
        }

    }
    
    answer = methods[n][m];



    return answer;
}

File: 2021/2/10/3.py
#프로그래머스 정수 삼각형


def solution(triangle):
    answer = 0

    dp = [[] for _ in range(len(triangle))]
    for i in range(len(dp)) :
        for j in range(len(triangle[i])) :
            dp[i].append(0)

    dp[0] = triangle[0]

    for i in range(1,len(dp)) :
        for j in range(len(dp[i])) :
            if j == 0 :
                dp[i][j] = max(dp[i][j],dp[i-1][j]+triangle[i][j])
            elif j == i :
                dp[i][j] = max(dp[i][j], dp[i - 1][j-1] + triangle[i][j])
            else :
                dp[i][j] = max(dp[i][j], dp[i - 1][j-1] + triangle[i][j],dp[i - 1][j] + triangle[i][j])


    answer = max(dp[-1])


    return answer